File: src/main/java/com/raoulvdberge/refinedstorage/api/autocrafting/craftingmonitor/ICraftingMonitorElementList.java
package com.raoulvdberge.refinedstorage.api.autocrafting.craftingmonitor;

import java.util.List;

public interface ICraftingMonitorElementList {
    /**
     * Directly add to the underlying list without trying to merge.
     *
     * @param element the {@link ICraftingMonitorElement}
     */
    void directAdd(ICraftingMonitorElement element);
    /**
     * Add a element to the Storage list, similar elements will be merged.
     * A {@link #commit()} will stop any following adds to be merged with previous ones.
     *
     * @param element the {@link ICraftingMonitorElement}
     */
    void addStorage(ICraftingMonitorElement element);
    /**
     * Add a element to the Processing or Crafting list, similar elements will be merged.
     * A {@link #commit()} will stop any following adds to be merged with previous ones.
     *
     * @param element the {@link ICraftingMonitorElement}
     * @param isProcessing wether to add to the processing list or the crafting list
     */
    void add(ICraftingMonitorElement element, boolean isProcessing);

    /**
     * Add a element to the list, similar elements will be merged.
     * A {@link #commit()} will stop any following adds to be merged with previous ones.
     *
     * @param element the {@link ICraftingMonitorElement}
     */
    void add(ICraftingMonitorElement element);

    /**
     * Finishes a current merge operation.
     */
    void commit();

    /**
     * Gets all the elements in the list.
     * This also commits the last changes.
     *
     * @return the current list of elements
     */
    List<ICraftingMonitorElement> getElements();
}

File: src/main/kotlin/com/refinedmods/refinedstorage/api/autocrafting/craftingmonitor/ICraftingMonitorElementList.kt
package com.refinedmods.refinedstorage.api.autocrafting.craftingmonitor


interface ICraftingMonitorElementList {
    /**
     * Directly add to the underlying list without trying to merge.
     *
     * @param element the [ICraftingMonitorElement]
     */
    fun directAdd(element: ICraftingMonitorElement)

    /**
     * Add a element to the Storage list, similar elements will be merged.
     * A [.commit] will stop any following adds to be merged with previous ones.
     *
     * @param element the [ICraftingMonitorElement]
     */
    fun addStorage(element: ICraftingMonitorElement)

    /**
     * Add a element to the Processing or Crafting list, similar elements will be merged.
     * A [.commit] will stop any following adds to be merged with previous ones.
     *
     * @param element      the [ICraftingMonitorElement]
     * @param isProcessing whether to add to the processing list or the crafting list
     */
    fun add(element: ICraftingMonitorElement, isProcessing: Boolean)

    /**
     * Add a element to the list, similar elements will be merged.
     * A [.commit] will stop any following adds to be merged with previous ones.
     *
     * @param element the [ICraftingMonitorElement]
     */
    fun add(element: ICraftingMonitorElement)

    /**
     * Finishes a current merge operation.
     */
    fun commit()

    /**
     * Gets all the elements in the list.
     * This also commits the last changes.
     *
     * @return the current list of elements
     */
    fun getElements(): List<ICraftingMonitorElement>
}